File: jetpackmvvm/src/main/java/com/zxn/mvvm/view/BaseFragment.java
package com.zxn.mvvm.view;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;

import com.zxn.mvvm.presenter.BasePresenter;
import com.zxn.mvvm.presenter.CreatePresenter;
import com.zxn.mvvm.presenter.IView;

import org.greenrobot.eventbus.EventBus;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Created by zxn on 2019/3/28.
 */
public abstract class BaseFragment<P extends BasePresenter> extends Fragment /*SimpleImmersionFragment*/ implements IView {

    private String TAG = "BaseFragment";

    protected P mPresenter;
    protected View mRootView;
    private Unbinder mUnbinder;
    protected String mPageTitle = "";
    protected FragmentActivity mContext;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (createPresenter() != null) {
            mPresenter = createPresenter();
            mPresenter.attachView(this);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (mRootView == null) {
            mRootView = getLayoutResId() <= 0
                    ? super.onCreateView(inflater, container, savedInstanceState)
                    : inflater.inflate(getLayoutResId(), container, false);
        } else {
            ViewGroup parent = (ViewGroup) mRootView.getParent();
            if (parent != null) {
                parent.removeView(mRootView);
            }
        }
        mUnbinder = ButterKnife.bind(this, mRootView);
        return mRootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mContext =  getActivity();
//        isPrepared = true;
        if (usedEventBus())
            regEventBus();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mUnbinder != null)
            mUnbinder.unbind();
        unRegEventBus();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mPresenter != null) {
            mPresenter.detachView();
        }
    }

    protected abstract int getLayoutResId();
    protected P createPresenter() {
        CreatePresenter annotation = getClass().getAnnotation(CreatePresenter.class);
        Class<P> pClass = null;
        if (annotation != null) {
            pClass = (Class<P>) annotation.value();
        }
        try {
            return pClass.newInstance();
        } catch (Exception e) {
//            throw new RuntimeException("Presenter创建失败!，检查是否声明了@CreatePresenter(xx.class)注解");
            Log.i(TAG, "Presenter创建失败!，检查是否声明了@CreatePresenter(xx.class)注解");
            return null;
        }
    }


    public CharSequence getPageTitle() {
        return mPageTitle;
    }

    /**
     * 是否使用EventBus，如果需要使用子类重载此方法并返回true
     *
     * @return 使用启用
     */
    protected boolean usedEventBus() {
        return false;
    }

    protected void regEventBus() {
        if (!EventBus.getDefault().isRegistered(this))
            EventBus.getDefault().register(this);
    }

    protected void unRegEventBus() {
        if (EventBus.getDefault().isRegistered(this))
            EventBus.getDefault().unregister(this);
    }

    @Override
    public void showLoading() {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.showLoading();
        }
    }

    @Override
    public void showLoading(String msg) {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.showLoading(msg);
        }
    }

    @Override
    public void showLoading(int msgResId) {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.showLoading(msgResId);
        }
    }

    @Override
    public void showLoading(boolean cancelable) {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.showLoading(cancelable);
        }
    }

    @Override
    public void showLoading(String msg, boolean cancelable) {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.showLoading(msg, cancelable);
        }
    }

    @Override
    public void closeLoading() {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.closeLoading();
        }
    }

    @Override
    public void showToast(int msg) {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.showToast(msg);
        }
    }

    @Override
    public void showToast(String msg) {
        if (getActivity() instanceof BaseActivity) {
            BaseActivity activity = (BaseActivity) getActivity();
            activity.showToast(msg);
        }
    }

    /**
     * Call this when your activity is done and should be closed.  The
     * ActivityResult is propagated back to whoever launched you via
     * onActivityResult().
     */
    public void finish() {
        if (null != getActivity()) {
            getActivity().finish();
        }
    }

    //    @Override
//    public void onResume() {
//        super.onResume();
//        if (getUserVisibleHint()) {
//            setUserVisibleHint(true);
//        }
//    }
//    @Override
//    public void setUserVisibleHint(boolean isVisibleToUser) {
//        super.setUserVisibleHint(isVisibleToUser);
//        if (getUserVisibleHint()) {
//            isVisible = true;
//            lazyLoad();
//        } else {
//            isVisible = false;
//            onInvisible();
//        }
//    }

//    /**
//     * 懒加载
//     */
//    protected void lazyLoad() {
//        if (!isPrepared || !isVisible || !isFirst) {
//            return;
//        }
//        initData();
//        isFirst = false;
//    }

//    /**
//     * fragment被设置为不可见时调用
//     */
//    protected abstract void onInvisible();


//    /**
//     * 这里获取数据，刷新界面，此方法执行在initView后面
//     */
//    protected abstract void initData();

    //    protected boolean isVisible;
//    private boolean isPrepared;//标志位，标志已经初始化完成
//    private boolean isFirst = true;

}

File: jetpackmvvm/src/main/java/com/zxn/mvvm/view/BaseFragment.kt
package com.zxn.mvvm.view

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import com.gyf.immersionbar.ImmersionBar
import com.trello.rxlifecycle2.components.support.RxFragment
import com.zxn.mvvm.network.NetState
import com.zxn.mvvm.network.NetworkStateManager
import com.zxn.mvvm.viewmodel.BaseViewModel

/**
 * Updated by zxn on 2020/10/23.
 */
abstract class BaseFragment : RxFragment(), IBaseView, ILoadingView {

    //是否第一次加载
    var isFirst: Boolean = true
        private set
    private var mPageTitle = ""

    override var titleBar: View? = null
    override var usedImmersionBar: Boolean = false
    override var usedStatusBarDarkFont: Boolean = false
    override var usedEventBus: Boolean = false
    override lateinit var mContext: AppCompatActivity
    override var cancelable: Boolean = true
    override val layoutResId: Int = 0
    override val layoutRoot: View? by lazy {
        onCreateRootView()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context as AppCompatActivity
    }

    /**
     * 网络变化监听 子类重写
     */
    open fun onNetworkStateChanged(netState: NetState) {

    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? = if (layoutRoot != null) layoutRoot else (if (layoutResId <= 0) super.onCreateView(
        inflater,
        container,
        savedInstanceState
    )
    else inflater.inflate(layoutResId, container, false))


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isFirst = true

        createObserver()

        onInitView()

        initData()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (usedEventBus) registerEventBus(true)
        onInitImmersionBar()
    }

    override fun onDestroy() {
        registerEventBus(false)
        super.onDestroy()
    }

    override fun onHiddenChanged(hidden: Boolean) {
        view?.post {
            if (!hidden) {
                lazyLoadData()
            }
        }
    }

    val pageTitle: CharSequence = mPageTitle

    override fun showLoading(msg: String?) {
        if (activity is BaseActivity/*<*>*/) {
            val activity = activity as BaseActivity/*<*>?*/
            activity.showLoading(msg)
        }
    }

    override fun showLoading(msgResId: Int) {
        if (activity is BaseActivity/*<*>*/) {
            val activity = activity as BaseActivity/*<*>?*/
            activity.showLoading(msgResId)
        }
    }


    override fun closeLoading() {
        if (activity is BaseActivity/*<*>*/) {
            val activity = activity as BaseActivity/*<*>?*/
            activity.closeLoading()
        }
    }

    override fun showToast(msg: Int) {
        if (activity is BaseActivity/*<*>*/) {
            val activity = activity as BaseActivity/*<*>?*/
            activity.showToast(msg)
        }
    }

    override fun showToast(msg: String) {
        if (activity is BaseActivity/*<*>*/) {
            val activity = activity as BaseActivity/*<*>?*/
            activity.showToast(msg)
        }
    }

    protected open fun finish() {
        activity?.finish()
    }

    override fun onInitImmersionBar() {
        if (usedImmersionBar) {
            if (usedStatusBarDarkFont) {
                setStatusBarDarkFont()
            } else {
                ImmersionBar.with(this)
                    .titleBar(titleBar)
                    .init()
            }
        }
    }

    /**
     * 白色状态栏,黑色字体,黑色导航栏,解决了白色状态栏无法看见状态栏字体颜色问题
     */
    open fun setStatusBarDarkFont() {
        ImmersionBar.with(this)
            .statusBarDarkFont(true)
            .navigationBarDarkIcon(true)
            .titleBar(titleBar)
            .init()
    }


    /**
     * Fragment执行onCreate后触发的方法
     */
    open fun initData() {}

    /**
     * 当可见的时候加载数据.
     */
    abstract fun lazyLoadData()

    /**
     * 是否需要懒加载
     */
    private fun onVisible() {
        if (lifecycle.currentState == Lifecycle.State.STARTED && isFirst) {
            //等待view加载后触发懒加载
            view?.post {
                //在Fragment中，只有懒加载过了才能开启网络变化监听
                NetworkStateManager.instance.mNetworkStateCallback.observeInFragment(this) {
                    //不是首次订阅时调用方法，防止数据第一次监听错误
                    if (!isFirst) {
                        onNetworkStateChanged(it)
                    }
                }
                isFirst = false
            }
        }
    }

    /**
     * 将非该Fragment绑定的ViewModel添加 loading回调 防止出现请求时不显示 loading 弹窗bug
     * @param viewModels Array<out BaseViewModel>
     */
    protected fun addLoadingObserve(vararg viewModels: BaseViewModel<*>) {
        viewModels.forEach { viewModel ->
            //显示弹窗
            viewModel.loadingChange.showDialog.observeInFragment(this) {
                showLoading()
            }
            //关闭弹窗
            viewModel.loadingChange.dismissDialog.observeInFragment(this) {
                closeLoading()
            }
        }
    }


}


//    /**
//     * 注册ViewModel与View的契约UI回调事件
//     */
//    private fun registorUIChangeLiveDataCallBack() {
//        //加载对话框显示
//        mViewModel.getUC().getShowLoadingEvent().observe(this, {
//            //showLoadingUI(it[BaseViewModel.ParameterField.MSG].toString(), it[BaseViewModel.ParameterField.IS_CANCLE] as Boolean)
//            cancelable = it[BaseViewModel.ParameterField.IS_CANCLE] as Boolean
//            if (it[BaseViewModel.ParameterField.MSG] is String) {
//                showLoading(it[BaseViewModel.ParameterField.MSG] as String)
//            }
//            if (it[BaseViewModel.ParameterField.MSG] is Int) {
//                showLoading(it[BaseViewModel.ParameterField.MSG] as Int)
//            }
//        })
//        //加载对话框消失
//        mViewModel.getUC().getHideLoadingEvent().observe(this, {
//            closeLoading()
//        })
//        //Toast显示
//        mViewModel.getUC().getShowToastEvent().observe(this, {
//            if (it is String) {
//                showToast(it)
//            }
//            if (it is Int) {
//                showToast(it)
//            }
//        })
//        //跳入新页面
//        mViewModel.getUC().getStartActivityEvent()
//                .observe(this, {
//                    fun onChanged(@Nullable params: Map<String, Any>) {
//                        val clz = params[BaseViewModel.ParameterField.CLASS] as Class<*>
//                        val bundle = params[BaseViewModel.ParameterField.BUNDLE] as Bundle
//                        startActivity(clz, bundle)
//                    }
//                })
//        //关闭界面
//        mViewModel.getUC().getFinishEvent().observe(this, {
//            finish()
//        })
//关闭上一层
//        mViewModel.getUC().getOnBackPressedEvent().observe(this, {
//            onBackPressed()
//        })
//    }

//    /**
//     * 注册 UI 事件
//     */
//    private fun registorDefUIChange() {
//        mViewModel.loadingChange.showDialog.observeInFragment(this) {
//            showLoading(it)
//        }
//        mViewModel.loadingChange.dismissDialog.observeInFragment(this) {
//            closeLoading()
//        }
//    }